File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/StatsClient.swift
import Foundation

enum StatsAction: String {
    case downloaded
    case applied
    case downloadError = "download_error"
    case rollback
}

enum StatsClient {
    static func send(
        updateUrl: String,
        appId: String,
        platform: String,
        action: StatsAction,
        bundleVersion: String?,
        channel: String?,
        releaseId: String?,
        nativeBuild: String?,
        errorMessage: String?
    ) {
        let sanitizedBase = updateUrl.replacingOccurrences(
            of: "/+$",
            with: "",
            options: .regularExpression
        )

        guard let url = URL(string: "\(sanitizedBase)/stats") else {
            return
        }

        var payload: [String: Any] = [
            "platform": platform,
            "action": action.rawValue,
        ]
        if let bundleVersion {
            payload["bundleVersion"] = bundleVersion
        }
        if let channel, !channel.isEmpty {
            payload["channel"] = channel
        }
        if let releaseId, !releaseId.isEmpty {
            payload["releaseId"] = releaseId
        }
        if let nativeBuild {
            payload["nativeBuild"] = nativeBuild
        }
        if let errorMessage {
            payload["errorMessage"] = String(errorMessage.prefix(500))
        }

        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(appId, forHTTPHeaderField: "X-App-Id")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        request.timeoutInterval = 10

        // Fire and forget - don't block on response
        URLSession.shared.dataTask(with: request).resume()
    }
}

File: packages/capacitor-plugin/android/src/main/java/com/updatekit/updater/StatsClient.java
package com.updatekit.updater;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import org.json.JSONObject;

final class StatsClient {

  private static final ExecutorService executor = Executors.newSingleThreadExecutor();

  private StatsClient() {}

  static void send(
    String updateUrl,
    String appId,
    String platform,
    String action,
    String bundleVersion,
    String channel,
    String releaseId,
    String nativeBuild,
    String errorMessage
  ) {
    executor.execute(() -> {
      HttpURLConnection connection = null;
      try {
        String base = updateUrl.replaceAll("/+$", "");
        URL url = new URL(base + "/stats");

        JSONObject payload = new JSONObject();
        payload.put("platform", platform);
        payload.put("action", action);
        if (bundleVersion != null) {
          payload.put("bundleVersion", bundleVersion);
        }
        if (channel != null && !channel.isEmpty()) {
          payload.put("channel", channel);
        }
        if (releaseId != null && !releaseId.isEmpty()) {
          payload.put("releaseId", releaseId);
        }
        if (nativeBuild != null) {
          payload.put("nativeBuild", nativeBuild);
        }
        if (errorMessage != null) {
          String truncated =
            errorMessage.length() > 500 ? errorMessage.substring(0, 500) : errorMessage;
          payload.put("errorMessage", truncated);
        }

        byte[] body = payload.toString().getBytes(StandardCharsets.UTF_8);

        connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("X-App-Id", appId);
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setConnectTimeout(10_000);
        connection.setReadTimeout(10_000);
        connection.setDoOutput(true);

        try (OutputStream output = connection.getOutputStream()) {
          output.write(body);
        }

        // Fire and forget - just trigger the request
        connection.getResponseCode();
      } catch (Exception ignored) {
        // Stats are best-effort, don't fail on errors
      } finally {
        if (connection != null) {
          connection.disconnect();
        }
      }
    });
  }
}
